concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -1151,6 +1154,47 @@ handleContractUpdateV1 originAddr istance checkAndGetSender transferAmount recei
+                            if not (Set.member (instanceInitName iParams) (GSWasm.miExposedInit mia)) then
+                                go events =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig rrdCurrentState False entryBalance (WasmV1.Error (WasmV1.EnvFailure (WasmV1.UpgradeInvalidContractName imuModRef (instanceInitName iParams)))) Nothing)
+                            else do
+                                -- Charge for the lookup of the new module.
+                                let iSize = GSWasm.miModuleSize iface
+                                tickEnergy $ Cost.lookupModule iSize
+                                -- Now we carry out the upgrade.
+                                -- The upgrade must preserve the following properties:
+                                -- 1. Subsequent operations in the receive function must be executed via the 'old' artifact.
+                                -- I.e. If some code is followed by the upgrade function the this must be run on the existing artifact code.
+                                -- 2. Reentrant contract calls must be executed on the new artifact.
+                                -- In order to fulfill the two above properties we add the actual state change to the 'ChangeSet' so that
+                                -- the 'resume' will execute on the 'old' version of the artifact. This solves (1).
+                                -- In order to fulfil (2) we will need to lookup the 'ChangeSet' whenever we get interrupted (with a call) to see
+                                -- if the code should be run on the old or the new artifact i.e., if a pending upgrade exists in the
+                                -- in the 'ChangeSet' we will need to use the new module artifact. See 'getCurrentContractInstance'.
+                                -- If the upgrade or subsequent actions fails then the transaction must be rolled back.
+                                -- Finally 'commitChanges' will commit the changeset upon a succesfull transaction.                                
+                                runExceptT (handleContractUpgrade cref imuModRef mia) >>= \case
+                                    Left errCode -> do
+                                        go (resumeEvent False:interruptEvent:events) =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig rrdCurrentState False entryBalance (WasmV1.Error (WasmV1.EnvFailure errCode)) Nothing)
+                                    Right upgradeEvents -> do
+                                        newBalance <- getCurrentContractAmount Wasm.SV1 istance
+                                        -- Charge for updating the pointer in the instance to the new module.
+                                        -- TODO: This is rather conservative and could probably be cheaper.
+                                        tickEnergy Cost.initializeContractInstanceCreateCost
```

Comment:
This is also not the right cost I think. This is for creating the new instance.

What we want to account here is
- looking up the module
- storing the new instance parameters

So as a first approximation the cost should be `initializeBaseCost + initializeCreateCost + lookupModule ms`

We should also charge this no matter whether the upgrade succeeds or not.